### PR TITLE
[#49623] Access to project folders for non members and admins.

### PR DESCRIPTION
**Pull request by ba1ash**
https://community.openproject.org/work_packages/49623

### CODE REVIEW

File: modules/storages/app/services/storages/group_folder_properties_sync_service.rb [hunk 1]
```diff
@@ -50,11 +51,15 @@ def initialize(storage)
   # rubocop:disable Metrics/AbcSize
   def call
     # we have to flush state to be able to reuse the instace of the class
-    @nextcloud_usernames_used_in_openproject = Set.new
     @project_folder_ids_used_in_openproject = Set.new
     @file_ids = nil
     @folders_properties = nil
     @group_users = nil
+    @admin_tokens_query = OAuthClientToken.where(oauth_client: @storage.oauth_client,
+                                                 users: User.where(admin: true, status: 'active'))
```
Comment (akabiru) [inline modules/storages/app/services/storages/group_folder_properties_sync_service.rb]:
```suggestion
    @admin_tokens_query = OAuthClientToken.where(oauth_client: @storage.oauth_client,
                                                 users: User.admin.active)
```

Using scopes might be more complete here. Apparently, the `User.active` scope ensures we exclude  system users, deleted users and anonymous users- via `Principals::Scopes::Status`. Is this behaviour valid for this use case?

```ruby
[7] pry(main)> User.admin.active.to_sql
=> "SELECT \"users\".* FROM \"users\" WHERE \"users\".\"type\" IN ('User', 'DeletedUser', 'AnonymousUser', 'SystemUser') AND \"users\".\"admin\" = TRUE AND \"users\".\"type\" NOT IN ('SystemUser', 'AnonymousUser', 'DeletedUser') AND \"users\".\"status\" = 1"
[8] pry(main)> User.where(admin: true, status: 'active').to_sql
=> "SELECT \"users\".* FROM \"users\" WHERE \"users\".\"type\" IN ('User', 'DeletedUser', 'AnonymousUser', 'SystemUser') AND \"users\".\"admin\" = TRUE AND \"users\".\"status\" = 1"
```

File: modules/storages/app/services/storages/group_folder_properties_sync_service.rb [hunk 2]
```diff
@@ -104,7 +109,7 @@ def ensure_project_folder(project_storage:, project_folder_path:)
     else
       create_folder(path: project_folder_path, project_storage:) >> obtain_file_id >> save_file_id
     end
-    # local variable is not used due to possible update_columns call
+    # local variable `project_folder_id` is not used due to possible update_columns call
```
Comment (akabiru) [inline modules/storages/app/services/storages/group_folder_properties_sync_service.rb:112]: 👍🏾

File: modules/storages/app/services/storages/group_folder_properties_sync_service.rb [hunk 3]
```diff
@@ -196,19 +201,34 @@ def group_users
...
-                          users: { "#{@nextcloud_system_user}": ALL_PERMISSIONS },
-                          groups: { "#{@group}": NO_PERMISSIONS }
-                        }) do |token, permissions|
+    tokens_query = OAuthClientToken
+                 .where(oauth_client: @storage.oauth_client)
+                 .where.not(id: @admin_tokens_query)
+                 .includes(:user)
+    if !(project.public? && Role.non_member.permissions.intersect?(PERMISSIONS_KEYS))
+      tokens_query = tokens_query.where(users: project.users)
+    end
```
Comment (akabiru) [inline modules/storages/app/services/storages/group_folder_properties_sync_service.rb:213]: This made my head spin a bit 😅  WDYT about extracting this to a descriptive method? / adding a comment above

```ruby
    if project_is_private_and_non_members_cannot_access_files?
      tokens_query = tokens_query.where(users: project.users)
    end
    # ....
    
     def project_is_private_and_non_members_cannot_access_files?
        !(project.public? && Role.non_member.permissions.intersect?(PERMISSIONS_KEYS))
     end
```

**Review by akabiru — APPROVED** [review]
Looking great and works well on test! 👍🏾  A few remarks for your consideration

File: modules/storages/app/services/storages/group_folder_properties_sync_service.rb (hunk 1 above)
Comment (ba1ash) [reply modules/storages/app/services/storages/group_folder_properties_sync_service.rb]: Yes, it makes sense for sure. Thanks!

File: modules/storages/app/services/storages/group_folder_properties_sync_service.rb (hunk 3 above)
Comment (ba1ash) [reply modules/storages/app/services/storages/group_folder_properties_sync_service.rb:213]: Yeah, it is cryptic :smile: I'll fix it.

**Pull request merged**